### PR TITLE
Handle errors in dynamic service resolution

### PR DESCRIPTION
**Pull request by felipe-gdr**
Introduces better error handling in the dynamic service resolution process.

Consumers should avoid throwing exceptions in the service resolution hook, and instead return GraphQL errors with helpful messages and metadata.

The Nadel engine will bypass normal execution for the root node which caused the error and instead craft a result node containing the error message and other details about the field.

### CODE REVIEW

File: engine/src/test/groovy/graphql/nadel/e2e/NadelE2ETest.groovy [hunk 1]
```diff
@@ -1436,220 +1436,4 @@ class NadelE2ETest extends StrategyTestHelper {
         response == [commentRenamed: [authorId: "fred", author: [displayName: "Display name of Fred"]]]
         errors.size() == 0
     }
-
```
Comment (felipe-gdr) [inline engine/src/test/groovy/graphql/nadel/e2e/NadelE2ETest.groovy:1439]: Split the tests into their own class to keep things more organised

File: api/src/main/java/graphql/nadel/hooks/ServiceExecutionHooks.java [hunk 2]
```diff
@@ -53,11 +54,10 @@ default NewVariableValue visitArgumentValueInQuery(HooksVisitArgumentValueEnviro
      * Called to resolve the service that should be used to fetch data for a field that uses dynamic service resolution.
      *
      * @param services a list of all services registered on Nadel
-     * @param fieldName the name of the field
-     * @param arguments GraphQL arguments that were passed when querying the field
+     * @param executionStepInfo
      * @return the Service that should be used to fetch data for that field or an error that was raised when trying to resolve the service.
      */
-    default ServiceOrError resolveServiceForField(List<Service> services, String fieldName, Map<String, Object> arguments) {
+    default ServiceOrError resolveServiceForField(List<Service> services, ExecutionStepInfo executionStepInfo) {
```
Comment (felipe-gdr) [inline api/src/main/java/graphql/nadel/hooks/ServiceExecutionHooks.java:60]: Getting a `ExecutionStepInfo` which contains everything the hook method will need to create a `GraphQLError`.

File: engine/src/test/groovy/graphql/nadel/e2e/NadelE2EDynamicServiceResolutionTest.groovy [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+import graphql.*
+import graphql.execution.AbortExecutionException
+import graphql.execution.ExecutionId
+import graphql.execution.ExecutionIdProvider
+import graphql.execution.ExecutionStepInfo
+import graphql.execution.instrumentation.InstrumentationContext
+import graphql.execution.instrumentation.InstrumentationState
+import graphql.execution.instrumentation.SimpleInstrumentationContext
+import graphql.language.SourceLocation
+import graphql.nadel.*
```
Comment (andimarek) [inline engine/src/test/groovy/graphql/nadel/e2e/NadelE2EDynamicServiceResolutionTest.groovy:12]: can you change your formatter to never use * imports? Have a look at the GraphQL formatter in the GJ repo.

File: engine/src/test/groovy/graphql/nadel/e2e/NadelE2EDynamicServiceResolutionTest.groovy [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+import java.util.concurrent.CompletionException
+
+import static graphql.language.AstPrinter.printAstCompact
+import static graphql.nadel.NadelEngine.newNadel
+import static graphql.nadel.NadelExecutionInput.newNadelExecutionInput
+import static graphql.nadel.engine.testutils.TestUtil.typeDefinitions
+import static java.lang.String.format
+import static java.util.concurrent.CompletableFuture.completedFuture
+
+class NadelE2EDynamicServiceResolutionTest extends StrategyTestHelper {
```
Comment (andimarek) [inline engine/src/test/groovy/graphql/nadel/e2e/NadelE2EDynamicServiceResolutionTest.groovy:43]: We should add at least on GraphQLE2ETest and not only StrategyTests.

File: engine/src/test/groovy/graphql/nadel/e2e/NadelE2EDynamicServiceResolutionTest.groovy [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+                return new ServiceOrError(services.stream().filter({ service -> (service.getName() == "RepoService") }).findFirst().get(), null)
+            }
+
+            if(idArgument.toString().contains("issue")) {
+                return new ServiceOrError(services.stream().filter({ service -> (service.getName() == "IssueService") }).findFirst().get(), null)
+            }
+
+            return new ServiceOrError(
+                    null,
+                    new GraphQLError() {
```
Comment (andimarek) [inline engine/src/test/groovy/graphql/nadel/e2e/NadelE2EDynamicServiceResolutionTest.groovy:104]: we have a GraphQLErrorBuilder in GJ which helps you to create errors.

**Review by andimarek — CHANGES_REQUESTED** [review]

File: engine/src/test/groovy/graphql/nadel/e2e/NadelE2ETest.groovy [hunk 6]
```diff
@@ -1348,308 +1349,4 @@ class NadelE2ETest extends StrategyTestHelper {
 
     }
 
-    def '''bugfix: check that a field that is used as an input for a hydration is not dropped from the response if it's
```
Comment (felipe-gdr) [inline engine/src/test/groovy/graphql/nadel/e2e/NadelE2ETest.groovy:1351]: Moved this test to the Strategy test class

**Review by andimarek — APPROVED** [review]
Great work!

**Pull request merged**